consult.el

```diff
@@ -1102,46 +1102,81 @@ for which the command history is used."
    ;; If pressing "C-x M-:", i.e., `repeat-complex-command',
    ;; we are instead querying the `command-history' and get a full s-expression.
    ((eq last-command 'repeat-complex-command)
-    (delete-dups (mapcar #'prin1-to-string command-history)))
+    (cons (delete-dups (mapcar #'prin1-to-string command-history))
+          'minibuffer-history-position))
    ;; In the minibuffer we use the current minibuffer history,
    ;; which can be configured by setting `minibuffer-history-variable'.
    ((minibufferp)
-    (if (fboundp 'minibuffer-history-value)
-        (minibuffer-history-value) ;; Emacs 27 only
-      (symbol-value minibuffer-history-variable)))
+    (cons (if (fboundp 'minibuffer-history-value)
+              (minibuffer-history-value) ;; Emacs 27 only
+            (symbol-value minibuffer-history-variable))
+          'minibuffer-history-position))
    ;; Otherwise we use a mode-specific history, see `consult-mode-histories'.
-   (t (when-let (history
-                 (or (seq-find (lambda (ring)
-                                 (and (derived-mode-p (car ring))
-                                      (boundp (cdr ring))))
-                               consult-mode-histories)
-                     (user-error
-                      "No history configured for `%s', see `consult-mode-histories'"
-                      major-mode)))
-        (symbol-value (cdr history))))))
+   (t (cl-destructuring-bind (history index-var)
+          (or (cdr (seq-find (lambda (ring)
+                               (and (derived-mode-p (car ring))
+                                    (boundp (cadr ring))))
+                             consult-mode-histories))
+              (user-error
+               "No history configured for `%s', see `consult-mode-histories'"
+               major-mode))
+        (cons (symbol-value history)
+              index-var)))))
+
+(defun consult--label-and-delete-dups (entries)
+  "Label ENTRIES with history indices.
+
+If there are duplicates, the most recent occurrence of the
+duplicate gets labeled with its index, and the others are not
+shown."
+  (let ((ht (make-hash-table :test 'equal))
+        (idx 0)
+        entry
+        accum)
+    (while (setq entry (pop entries))
+      (unless (gethash entry ht)
+        (puthash entry t ht)
+        (push (cons entry (cons idx entry))
+              accum))
+      (cl-incf idx))
+    (nreverse accum)))
 
 (defun consult--history-elements (history)
   "Return elements from HISTORY.
 Can handle lists and rings."
-  (delete-dups (seq-copy (if (ring-p history)
-                             (ring-elements history)
-                           history))))
+  (consult--label-and-delete-dups (seq-copy (if (ring-p history)
+                                                (ring-elements history)
+                                              history))))
```

Comment:
What are you doing here. Are you creating an alist (history-item . index) and then filtering out the duplicate history-items? This looks pretty complicated...

Reply:
Yes, the alist is of the form `((string . (history-index . same-string-again)) ...)`

When I wrote this, I assumed that the string would have to be listed once in the "key" portion of the alist to define the display string in the minibuffer, and once again (`same-string-again` in the above example) in the "value" portion of the alist because of how `consult--lookup-candidate` is used. Is there a nice way of storing the string only once while using `consult--lookup-candidate`?

Reply:
The candidate list should have one of the following forms:

* Only strings (list "candidate1" "candidate2" ...)
* Alist (list (cons "candidate1" data1) ...), if you use the lookup function the data1 is returned. This could be the index in our case.

Reply:
I shied away from storing the alist as `(("candidate-string" . INDEX) ... )` since storing the index alone means we have to go back and use the relevant lookup function (such as `eshell-get-history`) once again. Do you prefer this latter approach?

Reply:
@leungbk `consult-kmacro` also uses (string . index) pairs. I would prefer if we can use the same format for the histories.

Reply:
I don't know why you prefer the more complicated format. Not that it matters here, but it is certainly more expensive to allocate more cons pairs instead of doing another single lookup in the end.

Reply:
OK, I changed it to your way. I hadn't realized my way would be slower.